Updates to response handling (#70)
microsoftgraph/msgraph-sdk-design: ResponseHandler.md

## Requirements

- Provide the ability to take a response payload from a successful request and deserialize it into a strong type.
- Successful requests that MAY return a response payload MUST also support an empty response body.
- Provide the ability to take a response payload from a failed request and deserialize it into a ErrorContent object.
- Responses with 4XX or 5XX should attempt to translate response body into ErrorContent, otherwise a default ErrorContent should be created.
- Calls to the native Http library should not throw ServiceExceptions.

> - Service Library calls should throw an exception with a ErrorContent object. Ideally a different exception should be thrown based on a) the call could not be made b) the server returned an error.

Reviewer: ```suggestion
- Service Library calls SHOULD throw an exception with a ErrorContent object. Ideally a different exception should be thrown based on a) the call could not be made b) the server returned an error. See [ErrorContent Requirements](https://github.com/microsoftgraph/msgraph-sdk-design/blob/master/content/ErrorContent.md#requirements) on recommended exception names.
```
I'm a fan of RFC2119

> - Service Library calls should throw an exception with a ErrorContent object. Ideally a different exception should be thrown based on a) the call could not be made b) the server indicated that the client made an error c) the server made an error.

Author: ```suggestion
- Service Library calls should throw an exception with a ErrorContent object. Ideally a different exception should be thrown based on a) the call could not be made b) the server returned an error.
```

|POST | 200 | According to RFC 7231, the body contains the results or status of the action. In theory, this could be the created resource representation.  However, it is debatable whether it would be better to return a 201 with the actual response body.  |
|POST | 201 | According to RFC 7231, if this returns a body, it should contain a link and information about newly created object, not the object itself. Some APIs do return the created object. |
|POST | 204 | This MUST not return a body. |
|PATCH | 200 | According to RFC 7231, this should return the updated body. In practice, that isn't always the case.|
|PATCH | 201 | Same as POST.  |
|PATCH | 204 | This MUST not return a body. |
|PUT | 200 | According to RFC 7231, responses to PUTs are not cacheable, indicating that it is not expected to return the response body. However, returning the updated response does not seem to be prohibited. |
|PUT | 201 | Same rules as POST. |
|PUT | 204 | This MUST not return a body. |
|DELETE | 200 | The response body, if present should describe the status of the operation. In theory this could be a representation of the resource that was deleted. |
|DELETE | 204 | This MUST not return a body. |

> |GET | 200 | This MUST return a body that represents the identified resource. If should a representation doesn't exist, then the response SHOULD be a 404. |

Reviewer 2: nit:
```suggestion
|GET | 200 | This MUST return a body that represents the identified resource. If a representation doesn't exist, then the response SHOULD be a 404. |
```
